k2/csrc/thread_pool.cu

```diff
@@ -0,0 +1,94 @@
+/**
+ * @brief Thread pool.
+ *
+ *
+ * @copyright
+ * Copyright (c)  2020  Xiaomi Corporation (authors: Fangjun Kuang)
+ *
+ * @copyright
+ * See LICENSE for clarification regarding multiple authors
+ */
+
+#include <utility>
+
+#include "k2/csrc/thread_pool.h"
+
+namespace k2 {
+
+static int32_t GetDefaultNumThreads() {
+  int num_threads = std::thread::hardware_concurrency();
+  if (num_threads == 0) num_threads = 1;
+  return num_threads;
+}
+
+ThreadPool::ThreadPool(int32_t num_threads)
+    : threads_(num_threads > 0 ? num_threads : GetDefaultNumThreads()) {
+  for (auto &thread : threads_)
+    thread = std::thread([this]() { this->ProcessTasks(); });
+}
+
+ThreadPool::~ThreadPool() {
+  {
+    std::lock_guard<std::mutex> lock(mutex_);
+    keep_running_ = false;
+    not_empty_cond_.notify_all();  // wake up all threads in the pool
+  }
+  for (auto &thread : threads_) thread.join();
+}
+
+void ThreadPool::WaitAllTasksFinished() {
+  std::unique_lock<std::mutex> lock(mutex_);
+  while (!finished_) {
+    // wait for the `empty_cond_` condition.
+    empty_cond_.wait(lock);
+  }
+}
+
+void ThreadPool::ProcessTasks() {
+  std::unique_lock<std::mutex> lock(mutex_);
+  while (keep_running_) {
+    while (tasks_.empty() && keep_running_) {
+      // Wait for the `not_empty_` condition.
+      // RunTask() or the destructor will signal it.
+      not_empty_cond_.wait(lock);
```

Comment:
@csukuangfj , I think there's some performace issue, you use lock with  `mutex_` for all variables, e.g. not_empty_cond_, task, empty_cond, it's not ineffcient. I suggest here we can add an implemented of `SafeQueue` or class, it contains a separate `mutex` in it, so `task_.empty()`, `front`, `pop` will not lock `mutex_` here.

Reply:
> I think there's some performace issue, you use lock with mutex_ for all variables

All member variables are shared among threads, do you have any proposal modifying them without a lock?

> not_empty_cond_, task, empty_cond, it's not ineffcient

How would you use condition variables without holding a lock?

Reply:
I mean, they should not share the same mutex, if they shares, when other thread of `ProcessTask` try to use that `mutex`at the begining,  but it may be used by another thread that is doing `task_.pop()`, it's not reasonable. You should use a separate mutex for `task_`. the pinciple is: don't mix all non-direct-related variables shares a mutex, here `non_empty_cond_` and `task_` are two non-direct-related variables

Reply:
> if they shares, when other thread of ProcessTask try to use that mutex at the begining, but it may be used by another thread that is doing task_.pop(), it's not reasonable

Could you explain why it is not reasonable?  

How would you use a mutex to protect the condition variable `non_empty_cond_` if it needs protection?

Do we need a separate mutex for `keep_running` and a `mutex` for `finished_`?

As the region protected by the mutex is made as small as possible. I doubt it would be a performance bottleneck here.